src/Nethermind/Nethermind.JsonRpc/Modules/Subscribe/NewHeadSubscription.cs

```diff
@@ -0,0 +1,76 @@
+//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System.Threading.Tasks;
+using Nethermind.Blockchain;
+using Nethermind.JsonRpc.Modules.Eth;
+using Nethermind.Logging;
+
+namespace Nethermind.JsonRpc.Modules.Subscribe
+{
+    public class NewHeadSubscription : Subscription
+    {
+        private readonly IBlockTree _blockTree;
+        private readonly ILogger _logger;
+
+        public NewHeadSubscription(IBlockTree blockTree, ILogManager logManager)
+        {
+            _blockTree = blockTree;
+            _logger = logManager.GetClassLogger();
+            
+            _blockTree.NewHeadBlock += OnNewHeadBlock;
+            if(_logger.IsTrace) _logger.Trace($"NewHeads subscription {Id} will track NewHeadBlocks");
+        }
+        
+        private void OnNewHeadBlock(object? sender, BlockEventArgs e)
+        {
+            Task.Run(() =>
+            {
+                JsonRpcResult result =
+                    JsonRpcResult.Single(
+                        new JsonRpcSubscriptionResponse()
+                        {
+                            MethodName = "eth_subscribe",
+                            Params = new JsonRpcSubscriptionResult()
+                            {
+                                Result = new BlockForRpc(e.Block, false),
+                                Subscription = Id
+                            }
+                        }, default);
+
+                JsonRpcDuplexClient.SendJsonRpcResult(result);
+                if(_logger.IsTrace) _logger.Trace($"NewHeads subscription {Id} printed NewHeadBlock");
+            }).ContinueWith(
+                t =>
+                    t.Exception?.Handle(ex =>
+                    {
+                        if (_logger.IsDebug) _logger.Debug($"NewHeads subscription {Id}: Failed Task.Run after NewHeadBlock event.");
```

Comment:
Debug will be lost in normal flow, maybe at least info?

Reply:
this is always tricky - I ten to use DEBUG/ERROR for things like that, otherwise someone can flog our logs?

Reply:
*flood